src/System.Management.Automation/engine/NativeCommandProcessor.cs

```diff
@@ -130,6 +131,125 @@ internal ProcessOutputObject(object data, MinishellStream stream)
+        }
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="NativeCommandExitException"/> class.
+        /// </summary>
+        public NativeCommandExitException() : base() { }
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="NativeCommandExitException"/> class with a specified error message.
+        /// </summary>
+        /// <param name="message">
+        /// A localized error message.
+        /// </param>
+        public NativeCommandExitException(string message) : base(message) { }
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="NativeCommandExitException"/> class with a specified error message
+        /// and a reference to the inner exception that is the cause of this exception.
+        /// </summary>
+        /// <param name="message">
+        /// Localized error message.
+        /// </param>
+        /// <param name="innerException">
+        /// Inner exception.
+        /// </param>
+        public NativeCommandExitException(string message, Exception innerException) : base(message, innerException) { }
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="NativeCommandExitException"/> class with serialized data.
+        /// </summary>
+        /// <param name="info"></param>
+        /// <param name="context"></param>
+        protected NativeCommandExitException(SerializationInfo info, StreamingContext context)
+            : base(info, context)
+        {
+            if (info == null)
+            {
+                throw new PSArgumentNullException(nameof(info));
+            }
+
+            Path = info.GetString(nameof(Path));
+            ExitCode = info.GetInt32(nameof(ExitCode));
+            ProcessId = info.GetInt32(nameof(ProcessId));
+        }
+
+        #endregion Constructors
+
+        /// <summary>
+        /// Serializes the exception data.
+        /// </summary>
+        /// <param name="info">Serialization information.</param>
+        /// <param name="context">Streaming context.</param>
+        public override void GetObjectData(SerializationInfo info, StreamingContext context)
+        {
+            if (info == null)
+            {
+                throw new PSArgumentNullException(nameof(info));
+            }
+
+            base.GetObjectData(info, context);
+
+            info.AddValue(nameof(Path), Path);
+            info.AddValue(nameof(ExitCode), ExitCode);
+            info.AddValue(nameof(ProcessId), ProcessId);
+        }
+
+        /// <summary>
+        /// Gets the path of the native command.
+        /// </summary>
+        public string? Path { get; }
+
+        /// <summary>
+        /// Gets the exit code returned by the native command.
+        /// </summary>
+        public int? ExitCode { get; }
+
+        /// <summary>
+        /// Gets the native command's process Id.
+        /// </summary>
+        public int? ProcessId { get; }
```

Comment:
Is it explicit design or typo to make them nullable?

Reply:
There are constructors that don't set these and by not setting them they take the misleading value of `0`, so I made them explicitly nullable

Reply:
I think it make no sense for ExitCode.

Reply:
If this exception is instantiated with a constructor that doesn't set the exit code, it will leave it with the default of `0`. That's a contradiction, since this exception should only be thrown when the exit code is non-zero.

Reply:
In the case there _must_ be a constructor to validate and set a correct ExitCode.

>  this exception should only be thrown when the exit code is non-zero.

We should not make this assumption because there is no standard for return codes. For example, some interactive menu utility might return `-1` - cancel, `0` - default choice, `1` - first choice.

Reply:
> We should not make this assumption because there is no standard for return codes.

Well the entire reasoning in RFC 277 is that we enable a configuration where we assume a convention and throw when exit codes are non-zero. We are re-hashing the discussions around that RFC here and I don't think it's productive.

@rkeithhill what are the reasons for the other constructors that don't require arguments to set the exit code and process ID? Are we required to keep them for some reason?

Reply:
> we assume a convention

Yes, we do. I agree with this as first step. But SCCM allows to configure a response to exit codes for years. I guess we will get such feedback and new API should be flexible to allow the enhancement.

Reply:
On the constructor note, I've found [this](https://docs.microsoft.com/en-us/dotnet/fundamentals/code-analysis/quality-rules/ca1032), which says we should provide those constructors. Not sure how to square that with exception properties we want to ensure are set.

> new API should be flexible to allow the enhancement.

Well the idea is that `$PSNativeCommandUseErrorActionPreference` is set to `$false` by default, and the error is only thrown when it's set to `$true`.

Reply:
Also, there's this guidance: https://docs.microsoft.com/en-us/dotnet/standard/exceptions/how-to-create-user-defined-exceptions

Reply:
I reviewed a lot of exceptions https://source.dot.net/#System.Net.Http/Interop.GssApiException.cs,13 and not found the pattern with nullable value. 

ProcessId = 0 is Idle on Windows and I think it is ok to use 0 as implicit default.
ExitCode = 0 - Exception HResult is not nullable. I think it is ok to use 0 as implicit default too. Also we could reuse HResult and remove ExitCode at all (but keep the name in constructor parameters)

-----
We should make the class **sealed**.